### ISSUE TITLE
Updated analysis: update TMB analysis

### ISSUE THREAD

**Issue by jharenza** [PediatricOpenTargets/ticket-tracker#250]
#### What analysis module should be updated and why?
`snv-callers`  https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/dev/analyses/snv-callers/scripts/03-calculate_tmb.R

#### What changes need to be made? Please provide enough detail for another participant to make the update.
We should update the TMB calculation for all new samples and set this up for incoming data to be integrated flawlessly. Perhaps we should break the TMB piece out of this module into a new module (?). 

#### What input data should be used? Which data were used in the version being updated?
- WGS/WXS data should be used
- The last version run was OpenPBTA v21 using OpenPBTA consensus MAF, so PBTA data also needs an update.
- This will require the generation/release of padded BED files from GMKF, TARGET, and PNOC which were used in SNV calling (@runjin326 can you collate this information in another ticket for @zhangb1 ?) 

#### When do you expect the revised analysis will be completed?
?

#### Who will complete the updated analysis?
@runjin326 or @ewafula ? cc @afarrel for input

**Comment by ewafula** [comment]
@jharenza, @runjin32, I can work to update this module. But will first spend a bit of time to understand it first.

**Comment by jharenza** [comment]
> The last version run was OpenPBTA v21 using OpenPBTA consensus MAF, so PBTA data also needs an update.

@runjin326 for this, I meant that the new PBTA samples have not yet been run through this module, so they would need TMB calculations as well.

I also think it makes sense, in addition to releasing all BED files, to create a bed-mapping.tsv file which lists the BS_ID and the padded beds  column-wise per variant caller. Note: there are different BED files per sample because different variant callers use different padding.

**Comment by jharenza** [comment]
Adding some relevant references we used previously. Module may need some tweaking here since in OpenPBTA all callers were not used, as they are being used here in consensus. 

https://pubmed.ncbi.nlm.nih.gov/32217756/

https://www.annalsofoncology.org/article/S0923-7534(21)04495-1/fulltext

**Comment by ewafula** [comment]
@jharenza, @runjin326, @zhangb1,  based on the `OpenPBTA`, only `strelka2` and `mutect2`callers are used for TMB calculation. Considering that we only have the consensus file for `OpenPedCan` (`snv-consensus-plus-hotspots.maf.tsv.gz`) and not cohort-specific files for each caller (`mutect2`, `strelka2`, `verdict`, and `lancet`), I was intending to filter the consensus file for variant calls that have at least `strelka2` and `mutect2` for `SNVs`  and at least `mutect2` for `MNVs` (`strelka2` does not call `MNVs` which need to be split to SNVs for TMB calculation).  What was the reason why only these two callers are utilized OPenPBTA? Should this apply to OT cohorts as well?

In the OpenPBTA `release-v21-20210820` the following bed files among others are available  and utilized in the `snv-callers` module which also performs TMB calculation for PBTA (also available in `OPenPedcan v10 release` except for files prepended with an asterisk): 
```
intersect_cds_lancet_strelka_mutect_WGS.bed
*intersect_cds_lancet_WXS.bed
intersect_strelka_mutect_WGS.bed
WGS.hg38.lancet.300bp_padded.beds
WGS.hg38.lancet.unpadded.bed
WGS.hg38.mutect2.vardict.unpadded.bed
WGS.hg38.strelka2.unpadded.bed
WGS.hg38.vardict.100bp_padded.bed
*WXS.hg38.100bp_padded.bed
*WXS.hg38.lancet.400bp_padded.bed
```
For the PBTA TMB calculation two beds files are utilized, `intersect_strelka_mutect_WGS.bed` (bedtools intersect between `WGS.hg38.strelka2.unpadded.bed` and `WGS.hg38.mutect2.vardict.unpadded.bed`) and `WXS.hg38.100bp_padded.bed`. Therefore, I am expecting that we should generate similar bed files for the `GMKF`, `PNOC`, `TARGET` and `PNOC` if we are going to adapt the OpenPBTA approach of TMB calculation for all the OT cohorts.

**Comment by runjin326** [comment]
So Bo sent me all the bed files and explained the logic behind calling - some summaries and thoughts here: 
1) I think the reason for only using the intersect of two callers in OpenPBTA is because  Lancet uses a different bed file when calling MAF (while mutect2, vardict and strelka2 pretty much just used the whole genome chromosome file). Also , mutect2 and vardict uses the same file so using the mutect2 one should suffice. Since we are calling 2/4 for consensus MAF, I think sticking with the OpenPBTA method should be good for all WGS samples.
    - We can then use this for all WGS samples, including those in PBTA (which includes PNOC_dataset_1) and GMKF.
2) For PNOC_dataset_2, a similar manifest file with sample - seq-center - bed file can be used to map for each sample - I will organize the file.
3) For PNOC_dataset_1 WXS, I have the 100bp padded bed file - a similar manifest with sample - seq-center - bed file will be organized. 
4) For TARGET, I have the manifest and bed file and can share through slack. 

I think as @jharenza mentioned, we can release a bed-file mapping file that will be used for TMB calculation - I will generate a draft version with the above description. Feedback on above methodology, @jharenza? (I also read the two papers you attached - both of them are studying WXS vs. panel so a bit different from WGS. For WXS, one of them uses consensus region for all TCGA WES samples and the other one uses a bed file from F1CDx assay. Not much information about WGS regions to use or how many nts to pad).

Also, you mentioned that different bed files needed to be padded differently - for PNOC dataset 2, 2nt was used and for PNOC dataset 1, 100bp was used. Is there a guideline somewhere to determine how many nucleotides to use?

**Comment by jharenza** [comment]
Hi @runjin326 - your methodology sounds good.

> for PNOC dataset 2, 2nt was used and for PNOC dataset 1, 100bp was used

I was under the impression that the nt used for padding WXS beds were based on the variant caller not the dataset. @migbro can you weigh in here as to why these paddings would differ?

**Comment by migbro** [comment]
Only thing I could think of is that in batch 1, they were accidentally aggressively padded and we noted that. Batch 2 I think "fixes" this, however the 2bp pad sounds like a choice an analyst made?

**Comment by runjin326** [comment]
@jharenza, looks like for OpenPBTA - the WXS region file is padded with 100bp and then used for TMB calculation. For PNOC (both dataset1 and 2) and TARGET samples (regardless of either WXS or panel) - the padded files that we have in hand are all 100bp padded (wondering what they are for). Should we maybe just use 100bp padded for all panel + WXS samples to be consistent? 

Another question is in OpenPBTA `intersect_strelka_mutect_WGS.bed` is generated by intersecting `WGS.hg38.strelka2.unpadded.bed` and `WGS.hg38.mutect2.vardict.unpadded.bed` while currently @zhangb1 mentioned we are using `wgs_canonical_calling_regions.hg38.bed` for mutect2 and vardict and `hg38_strelka.bed` for strelka - should we do our own intersect?

**Comment by ewafula** [comment]
@jharenza, @runjin32,  in the  OpenPBTA `snv_callers` module , only `strelka2` and `mutect2` calls are used for TMB calculation. Should we also ignore calls specific to only `lancet`, `VarDict`, and `VarDict,Lancet` in OT ?  Some tumor samples will be left out. In v10 17 samples, all TARGET NBL.
```
> df <- fread("snv-consensus-plus-hotspots.maf.tsv.gz", skip=1, select=c("Tumor_Sample_Barcode", "CAL"), showProgress = FALSE) %>% tibble()
> df %>% distinct(CAL)
# A tibble: 15 x 1
   CAL                            
   <chr>                          
 1 Strelka2,Mutect2,VarDict       
 2 Strelka2,Mutect2,VarDict,Lancet
 3 Strelka2,VarDict               
 4 Strelka2,Mutect2               
 5 Strelka2,Lancet                
 6 Mutect2,VarDict,Lancet         
 7 VarDict,Lancet                 
 8 Mutect2,VarDict                
 9 Mutect2,Lancet                 
10 Strelka2,VarDict,Lancet        
11 Strelka2,Mutect2,Lancet        
12 VarDict                        
13 Mutect2                        
14 Strelka2                       
15 Lancet                         
> df %>% filter(CAL %in% c("VarDict", "Lancet", "VarDict,Lancet"))
# A tibble: 207,873 x 2
   Tumor_Sample_Barcode CAL           
   <chr>                <chr>         
 1 BS_ZZJF26C4          VarDict,Lancet
 2 BS_ZZJF26C4          VarDict,Lancet
 3 BS_ZZJF26C4          VarDict,Lancet
 4 BS_ZZJF26C4          VarDict,Lancet
 5 BS_ZZJF26C4          VarDict,Lancet
 6 BS_ZZJF26C4          VarDict,Lancet
 7 BS_ZZJF26C4          VarDict,Lancet
 8 BS_ZZJF26C4          VarDict,Lancet
 9 BS_ZZJF26C4          VarDict,Lancet
10 BS_ZZJF26C4          VarDict,Lancet
# ... with 207,863 more rows
> vardict_lancet_only_samples <- df %>% filter(CAL %in% c("VarDict", "Lancet", "VarDict,Lancet"))  %>% distinct(Tumor_Sample_Barcode)
> all_other_samples <- df %>% filter(!CAL %in% c("VarDict", "Lancet", "VarDict,Lancet"))  %>% distinct(Tumor_Sample_Barcode)
> setdiff(vardict_lancet_only_samples, all_other_samples)
# A tibble: 17 x 1
   Tumor_Sample_Barcode    
   <chr>                   
 1 TARGET-30-PATKGB-01A-01D
 2 TARGET-30-PASEAR-01A-01D
 3 TARGET-30-PASTIJ-01A-01D
 4 TARGET-30-PASPIK-01A-01D
 5 TARGET-30-PASFCG-01A-01D
 6 TARGET-30-PASCRK-01A-01D
 7 TARGET-30-PARTCE-01A-01D
 8 TARGET-30-PASBJY-01A-01D
 9 TARGET-30-PASGHN-01A-01D
10 TARGET-30-PATFTN-01A-01D
11 TARGET-30-PATPJD-09A-01D
12 TARGET-30-PAULMT-01A-01D
13 TARGET-30-PAUZTF-01A-01D
14 TARGET-30-PASSGT-01A-01D
15 TARGET-30-PASDDP-01A-01D
16 TARGET-30-PARWBC-01A-01D
17 TARGET-30-PASGCD-01A-01D
```

**Comment by runjin326** [comment]
@ewafula, I think we want to do something different from OpenPBTA - and maybe take out the TMB calculation from the `snv_callers` module and then separate a TMB calculation module just for it. 
But here - I think the idea is we would use straight from the consensus MAF file that we included in the data release (regardless of what callers it used - as long as it is in the consensus file) and just count the number of mutations in that file for each sample. Maybe @jharenza can confirm.

**Comment by ewafula** [comment]
> @ewafula, I think we want to do something different from OpenPBTA - and maybe take out the TMB calculation from the `snv_callers` module and then separate a TMB calculation module just for it. But here - I think the idea is we would use straight from the consensus MAF file that we included in the data release (regardless of what callers it used - as long as it is in the consensus file) and just count the number of mutations in that file for each sample. Maybe @jharenza can confirm.

Thanks, @runjin326. I will proceed with all callers.

**Comment by jharenza** [comment]
Yes, we would do different than OpenPBTA as @runjin326 suggested. Interesting about 

> Some tumor samples will be left out. In v10 17 samples, all TARGET NBL.

I wonder if those are real calls or could be artifacts... cc @afarrel on this

**Comment by jharenza** [comment]
closed with #157
